lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,159 @@
+	}
+	if numBlocks == 0 {
+		return errors.New("PoSFeeEstimator.Init: numBlocks cannot be zero")
+	}
+	if numBlocks != uint64(len(pastBlocks)) {
+		return errors.New("PoSFeeEstimator.Init: numBlocks must equal the number of pastBlocks")
+	}
+	// Sort the past blocks by height just to be safe.
+	sortedPastBlocks := collections.SortSliceStable(pastBlocks, func(ii, jj *MsgDeSoBlock) bool {
+		return ii.Header.Height < jj.Header.Height
+	})
+	minHeight := sortedPastBlocks[0].Header.Height
+	if minHeight == 0 {
+		return errors.New("PoSFeeEstimator.Init: minHeight cannot be zero")
+	}
+	// TODO: Are there any other inputs to validate?
+	posFeeEstimator.mempoolFeeEstimatorMempool = mempool
+	posFeeEstimator.numBlocks = numBlocks
+
+	// TODO: Can we get away without setting the pastBlocksUtxoView? This would
+	// save us a badger instance that isn't doing anything.
+
+	// TODO: How do we know when to update global params on the pastBlocksMempool if they're updated on the original mempool?
+	// We only really need that global params entry to help us compute the fee buckets. It's on the caller to update it
+	// using
+	// Create a mempool we can use to estimate fees for past blocks.
+	posFeeEstimator.pastBlocksTransactionRegister = NewTransactionRegister(mempool.globalParams.Copy())
+
+	// Add all the txns from the past blocks to the new pastBlocksMempool.
+	for _, block := range sortedPastBlocks {
+		if err := posFeeEstimator.addBlock(block); err != nil {
+			return errors.Wrap(err, "PosFeeEstimator.Init: error adding block to pastBlocksMempool")
+		}
+	}
+	return nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) AddBlock(block *MsgDeSoBlock) error {
+	return posFeeEstimator.addBlock(block)
+}
+
+func (posFeeEstimator *PoSFeeEstimator) addBlock(block *MsgDeSoBlock) error {
+	// Add all transactions from the block to the pastBlocksTransactionRegister.
+	for _, txn := range block.Txns {
+		mtxn, err := NewMempoolTx(txn, block.Header.TstampNanoSecs, block.Header.Height)
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.AddBlock: error creating MempoolTx")
+		}
+		if err = posFeeEstimator.pastBlocksTransactionRegister.AddTransaction(mtxn); err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.AddBlock: error adding txn to pastBlocksTransactionRegister")
+		}
+	}
+	posFeeEstimator.cachedBlocks = append(posFeeEstimator.cachedBlocks, block)
+	// Sort the cached blocks by height & tstamp just to be safe.
+	posFeeEstimator.sortCachedBlocks()
+	if uint64(len(posFeeEstimator.cachedBlocks)) > posFeeEstimator.numBlocks {
+		// Remove the oldest block.
+		if err := posFeeEstimator.RemoveBlock(posFeeEstimator.cachedBlocks[0]); err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.AddBlock: error removing block from PoSFeeEstimator")
+		}
+	}
+	return nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) RemoveBlock(block *MsgDeSoBlock) error {
+	// Remove all transaction from the block from the pastBlocksTransactionRegister.
```

Comment:
Wrap these public functions with a read-write mutex
```suggestion
	posFeeEstimator.lock.Lock()
	defer posFeeEstimator.lock.Unlock()
	
	// Remove all transaction from the block from the pastBlocksTransactionRegister.
```